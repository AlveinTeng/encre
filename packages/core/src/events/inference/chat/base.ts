import { TiktokenModel } from 'js-tiktoken/lite';
import { BaseCache } from '../../../cache/base.js';
import { MemoryCache } from '../../../cache/index.js';
import { type CallableConfig } from '../../../record/callable.js';
import {
  AsyncCallError,
  AsyncCaller,
  baseFailedAttemptHandler,
  type AsyncCallerParams,
} from '../../../utils/asyncCaller.js';
import {
  encodingForModel,
  getNumTokens,
  getTiktokenModel,
} from '../../../utils/tokenizer.js';
import { BaseEvent, BaseEventParams } from '../../base.js';
import {
  type BaseMessageLike,
  BaseMessage,
} from '../../input/load/msgs/base.js';
import { convertMessageLikeToMessage, isMessageLike } from '../../input/load/msgs/utils.js';
import { BasePrompt } from '../../input/load/prompts/base.js';
import { ChatPrompt } from '../../input/load/prompts/chat.js';
import { StringPrompt } from '../../input/load/prompts/text.js';
import { Generation } from '../../output/provide/generation.js';
import { LLMResult } from '../../output/provide/llmresult.js';

export interface BaseLMCallOptions extends BaseEventParams {
  /**
   * Stop tokens in the LM call.
   * If not provided, default stop tokens are used for the LM.
   */
  stopWords?: string[];

  /**
   * Timeout for the LM cal in milliseconds;
   */
  timeout?: number;

  /**
   * Abort signal for the call.
   * If provided, the call will be aborted when the signal is aborted.
   */
  signal?: AbortSignal;
}

export type BaseLMInput = BasePrompt | BaseMessageLike[] | string;

export interface BaseLMParams extends AsyncCallerParams, BaseEventParams {
  cache?: BaseCache | boolean;
}

/**
 * Abstract class representing the BaseLM (Base Language Model).
 * This class offers core functionalities for interfacing with a language model, including caching,
 * token counting, and transforming inputs to prompts.
 *
 * @template CallOutput - Represents the call output type. Default is unknown.
 * @template CallOptions - Represents the call options type. By default, it extends from {@link BaseLMCallOptions}.
 */
export abstract class BaseLM<
    CallOutput = unknown,
    CallOptions extends BaseLMCallOptions = BaseLMCallOptions,
  >
  extends BaseEvent<BaseLMInput, CallOutput, CallOptions>
  implements BaseLMParams
{
  /**
   * Represents the type for call options.
   */
  declare CallOptions: CallOptions;

  /**
   * Instance responsible for making asynchronous calls.
   */
  caller: AsyncCaller;

  /**
   * Cache instance to store and retrieve results for given prompts.
   */
  cache?: BaseCache;

  /**
   * Constructor for the BaseLM class.
   * @param {BaseLMParams} params - Parameters to initialize the base language model.
   */
  constructor({ callbacks, ...params }: BaseLMParams) {
    super({ callbacks, ...params });

    if (typeof params.cache === 'object') {
      this.cache = params.cache;
    } else if (params.cache) {
      this.cache = MemoryCache.global();
    } else {
      this.cache = undefined;
    }

    this.caller = new AsyncCaller(
      (params ?? {}) && { onFailedAttempt: this._failedAttemptHandler }
    );
  }

  /**
   * Abstract method to get the LLM type. Must be implemented by subclasses.
   * @returns {string} The type of LLM.
   */
  abstract _llmType(): string;

  /**
   * Abstract method to get the model type. Must be implemented by subclasses.
   * @returns {string} The type of the model.
   */
  abstract _modelType(): string;

  /**
   * Abstract method to invoke the language model with a given input and options.
   * @param input - The input for the language model.
   * @param options - Optional call options.
   * @returns {Promise<LLMResult>} The output llm result from the language model.
   */
  abstract invoke(
    input: BaseLMInput,
    options?: Partial<CallOptions>
  ): Promise<CallOutput>;

  /**
   * Abstract method to provide the core logic to interface with the language model,
   * handling both cached and uncached predictions.
   * @param {unknown} input - A given input to the language model, can be a prompt or an array of messages.
   * @param {string[] | CallOptions} [options] - Optional call options or an array of stop words.
   * @param {any} [callbacks] - Optional callbacks.
   * @returns {Promise<LLMResult>} The result from the language model.
   */
  abstract provide(
    input: unknown,
    options?: string[] | CallOptions,
    callbacks?: any
  ): Promise<LLMResult>;

  /**
   * Converts a given {@link BaseLMInput} to a {@link BasePrompt}.
   * @param {BaseLMInput} input - The input for the language model.
   * @returns {BasePrompt} The corresponding prompt.
   */
  protected static _convertInputToPrompt(input: BaseLMInput): BasePrompt {
    if (typeof input === 'string') {
      return new StringPrompt(input);
    } else if (Array.isArray(input)) {
      return new ChatPrompt(
        input.map((msg: BaseMessageLike) => convertMessageLikeToMessage(msg))
      );
    }

    return input;
  }

  /**
   * Method to identify additional parameters specific to implementations.
   * @returns {Record<string, any>} A record of identified parameters.
   */
  protected _identifyParams(): Record<string, any> {
    return {};
  }

  /**
   * Constructs a string key based on the given call options for caching purposes.
   * @param {CallOptions} callOptions - The call options.
   * @returns {string} The generated key.
   */
  protected _getLLMStrKey(callOptions: CallOptions): string {
    const params: Record<string, any> = {
      ...this._identifyParams(),
      ...callOptions,
      _type: this._llmType(),
      _model: this._modelType(),
    };

    const nonEmptyParams: Record<string, any> = Object.entries(params).filter(
      ([_, v]) => v !== undefined
    );

    const llmStrKey: string = nonEmptyParams
      .map(([k, v]) => `${k}:${JSON.stringify(v)}`)
      .sort()
      .join(',');

    return llmStrKey;
  }

  protected _failedAttemptHandler(e: Error) {
    baseFailedAttemptHandler(e as AsyncCallError);
  }
}

export interface BaseLLMCallOptions extends BaseLMCallOptions {}

export interface BaseLLMParams extends BaseLMParams {}

/**
 * Abstract class representing the BaseLLM. This class provides core functionalities
 * to interface with a language model, potentially allowing for both cached and uncached predictions.
 *
 * @template CallOptions - Represents the call options type. By default, it extends from {@link BaseLLMCallOptions}.
 */
export abstract class BaseLLM<
  CallOptions extends BaseLLMCallOptions = BaseLLMCallOptions,
> extends BaseLM<LLMResult, CallOptions> {
  /**
   * Represents the type for call options without some specified properties.
   */
  declare SerializedCallOptions: Omit<
    CallOptions,
    keyof CallableConfig & 'timeout'
  >;

  /**
   * A predefined namespace array to identify the type of language model and other related namespaces.
   */
  _eventNamespace(): string[] {
    return ['inference', 'chat', 'llms', this._llmType()];
  }

  /**
   * Constructor for the BaseLLM class.
   * @param fields - Parameters to initialize the base language model.
   */
  constructor(fields: BaseLLMParams) {
    super(fields);
  }

  /**
   * Returns the type of the model.
   * @returns {string} The type of the model.
   */
  _modelType(): string {
    return 'base_llm';
  }

  abstract getParams(
    options?: this['SerializedCallOptions']
  ): Record<string, unknown>;

  /**
   * Invokes the language model with a given input and options.
   * @param input - The input for the language model.
   * @param options - Optional call options.
   * @returns {Promise<LLMResult>} The output llm result from the language model.
   */
  async invoke(input: BaseLMInput, options?: CallOptions): Promise<LLMResult> {
    const prompt: BasePrompt = BaseLLM._convertInputToPrompt(input);
    const result: LLMResult = await this.provide(
      prompt.toString(),
      options,
      options?.callbacks
    );

    return result;
  }

  /**
   * Provides the core logic to interface with the language model, handling both cached and uncached predictions.
   * @param {string} prompt - A prompt.
   * @param options - Optional call options or an array of stop words.
   * @param callbacks - Optional callbacks.
   * @returns {Promise<LLMResult>} The result from the language model.
   */
  async provide(
    prompt: string,
    options?: CallOptions | string[],
    callbacks?: any
  ): Promise<LLMResult> {
    if (Array.isArray(prompt) || !(typeof prompt === 'string')) {
      throw new Error("Argument 'prompt' must be string");
    }

    let parsedOptions: CallOptions | undefined;
    if (Array.isArray(options)) {
      parsedOptions = { stopWords: options } as CallOptions;
    } else {
      parsedOptions = options;
    }

    const [callableOptions, serializedCallOptions] =
      this._splitCallableOptionsFromCallOptions(parsedOptions);
    callableOptions.callbacks = callableOptions.callbacks ?? callbacks;

    const llmStrKey: string = this._getLLMStrKey(serializedCallOptions);

    let generations: Generation[] | null | undefined = await this.cache?.lookup(
      [prompt, llmStrKey]
    );

    let llmOutput = {};
    if (generations === null || generations === undefined) {
      const llmResult: LLMResult = await this._provideUncached(
        prompt,
        serializedCallOptions
      );
      await this.cache?.update([prompt, llmStrKey], llmResult.generations);

      generations = llmResult.generations;
      llmOutput = llmResult.llmOutput ?? {};
    }

    return { generations, llmOutput };
  }

  /**
   * Abstract method that interfaces with the underlying language model. Must be implemented by subclasses.
   * @param {string} prompt - A prompt.
   * @param options - Call options.
   * @returns {Promise<LLMResult>} The result from the language model.
   */
  abstract _provide(
    prompt: string,
    options: this['SerializedCallOptions']
  ): Promise<LLMResult>;

  /**
   * Handles uncached prompts and calls the `_provide` method.
   * @param {string} prompt - A prompt.
   * @param serializedCallOptions - Serialized call options.
   * @returns {Promise<LLMResult>} The result from the language model.
   */
  protected async _provideUncached(
    prompt: string,
    serializedCallOptions: this['SerializedCallOptions']
  ): Promise<LLMResult> {
    let output: LLMResult;

    /*eslint no-useless-catch: "warn"*/
    try {
      output = await this._provide(prompt, serializedCallOptions);
    } catch (e) {
      throw e;
    }

    return output;
  }

  /**
   * Splits the provided call options into callable options and serialized options.
   * @param options - Call options.
   * @returns A tuple containing callable options and serialized options.
   */
  protected _splitCallableOptionsFromCallOptions(
    options?: Partial<CallOptions>
  ): [CallableConfig, this['SerializedCallOptions']] {
    const [callableOptions, serializedCallOptions] =
      super._splitCallableOptionsFromCallOptions(options);

    if (serializedCallOptions?.timeout && !serializedCallOptions.signal) {
      serializedCallOptions.signal = AbortSignal.timeout(
        serializedCallOptions.timeout
      );
    }

    return [
      callableOptions,
      serializedCallOptions as this['SerializedCallOptions'],
    ];
  }
}

export interface BaseChatLMCallOptions extends BaseLMCallOptions {}

export interface BaseChatLMParams extends BaseLMParams {}

/**
 * Abstract class representing the BaseChatLM. This class provides core functionalities
 * to interface with a language model, potentially allowing for both cached and uncached predictions.
 *
 * @template CallOptions - Represents the call options type. By default, it extends from {@link BaseChatLMCallOptions}.
 */
export abstract class BaseChatLM<
  CallOptions extends BaseChatLMCallOptions = BaseChatLMCallOptions,
> extends BaseLM<LLMResult, CallOptions> {
  /**
   * Represents the type for call options without some specified properties.
   */
  declare SerializedCallOptions: Omit<
    CallOptions,
    keyof CallableConfig & 'timeout'
  >;

  /**
   * A predefined namespace array to identify the type of language model and other related namespaces.
   */
  _eventNamespace(): string[] {
    return ['inference', 'chat', 'chatlms', this._llmType()];
  }

  constructor(fields: BaseChatLMParams) {
    super(fields);
  }

  /**
   * Returns the type of the model.
   * @returns {string} The type of the model.
   */
  _modelType(): string {
    return 'base_chatlm';
  }

  abstract getParams(
    options?: this['SerializedCallOptions']
  ): Record<string, unknown>;

  /**
   * Invokes the chat language model with a given input and options.
   * @param input - The input for the chat language model.
   * @param options - Optional call options.
   * @returns {Promise<LLMResult>} The output llm result from the language model.
   */
  async invoke(input: BaseLMInput, options?: CallOptions): Promise<LLMResult> {
    const prompt: BasePrompt = BaseChatLM._convertInputToPrompt(input);
    const result: LLMResult = await this.provide(
      prompt.toChatMessages(),
      options,
      options?.callbacks
    );

    return result;
  }

  /**
   * Provides the core logic to interface with the chat language model, handling both cached and uncached predictions.
   * @param {BaseMessageLike[]} messages - An array of messages.
   * @param options - Optional call options or an array of stop words.
   * @param callbacks - Optional callbacks.
   * @returns {Promise<LLMResult>} The result from the language model.
   */
  async provide(
    messages: BaseMessageLike[],
    options?: CallOptions | string[],
    callbacks?: any
  ): Promise<LLMResult> {
    if (!Array.isArray(messages)) {
      throw new Error("Argument 'messages' must be array.");
    }

    const baseMessages: BaseMessage[] = messages.map(
      convertMessageLikeToMessage
    );

    let parsedOptions: CallOptions | undefined;
    if (Array.isArray(options)) {
      parsedOptions = { stopWords: options } as CallOptions;
    } else {
      parsedOptions = options;
    }

    const [callableOptions, serializedCallOptions] =
      this._splitCallableOptionsFromCallOptions(parsedOptions);
    callableOptions.callbacks = callableOptions.callbacks ?? callbacks;

    const llmStrKey: string = this._getLLMStrKey(serializedCallOptions);

    const prompt: string =
      BaseChatLM._convertInputToPrompt(baseMessages).toString();
    let generations: Generation[] | null | undefined = await this.cache?.lookup(
      [prompt, llmStrKey]
    );

    let llmOutput = {};
    if (generations === null || generations === undefined) {
      const llmResult: LLMResult = await this._provideUncached(
        baseMessages,
        serializedCallOptions
      );
      await this.cache?.update([prompt, llmStrKey], llmResult.generations);

      generations = llmResult.generations;
      llmOutput = llmResult.llmOutput ?? {};
    }

    return { generations, llmOutput };
  }

  /**
   * Abstract method that interfaces with the underlying language model. Must be implemented by subclasses.
   * @param {BaseMessage[]} messages - An array of messages.
   * @param options - Call options.
   * @returns {Promise<LLMResult>} The result from the language model.
   */
  abstract _provide(
    messages: BaseMessage[],
    options: this['SerializedCallOptions']
  ): Promise<LLMResult>;

  /**
   * Handles uncached prompts and calls the `_provide` method.
   * @param {BaseMessage[]} messages - An array of messages.
   * @param serializedCallOptions - Serialized call options.
   * @returns {Promise<LLMResult>} The result from the language model.
   */
  protected async _provideUncached(
    messages: BaseMessage[],
    serializedCallOptions: this['SerializedCallOptions']
  ): Promise<LLMResult> {
    let output: LLMResult;

    /*eslint no-useless-catch: "warn"*/
    try {
      output = await this._provide(messages, serializedCallOptions);
    } catch (e) {
      throw e;
    }

    return output;
  }

  /**
   * Splits the provided call options into callable options and serialized options.
   * @param options - Call options.
   * @returns A tuple containing callable options and serialized options.
   */
  protected _splitCallableOptionsFromCallOptions(
    options?: Partial<CallOptions>
  ): [CallableConfig, this['SerializedCallOptions']] {
    const [callableOptions, serializedCallOptions] =
      super._splitCallableOptionsFromCallOptions(options);

    if (serializedCallOptions?.timeout && !serializedCallOptions.signal) {
      serializedCallOptions.signal = AbortSignal.timeout(
        serializedCallOptions.timeout
      );
    }

    return [
      callableOptions,
      serializedCallOptions as this['SerializedCallOptions'],
    ];
  }
}

export async function calculateMaxToken(
  prompt: string,
  modelName: TiktokenModel
): Promise<number> {
  const numTokens: number = await getNumTokens(prompt, modelName);

  const maxTokens = getModelContextSize(modelName);
  return maxTokens - numTokens;
}

export function getModelContextSize(modelName: string): number {
  switch (getTiktokenModel(modelName)) {
    case 'gpt-3.5-turbo-16k':
      return 16384;
    case 'gpt-3.5-turbo':
      return 4096;
    case 'gpt-4-32k':
      return 32768;
    case 'gpt-4':
      return 8192;
    case 'text-davinci-003':
      return 4096;
    case 'text-curie-001':
      return 2048;
    case 'text-babbage-001':
      return 2048;
    case 'text-ada-001':
      return 2048;
    case 'code-davinci-002':
      return 8000;
    case 'code-cushman-001':
      return 2048;
    default:
      return 4096;
  }
}

export function isLMInput(value: unknown): value is BaseLMInput {
  if (Array.isArray(value)) {
    return value.every((v) => isMessageLike(v));
  }

  return typeof value === 'string' || value instanceof BasePrompt;
}
