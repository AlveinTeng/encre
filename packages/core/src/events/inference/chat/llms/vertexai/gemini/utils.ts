import { Content, InlineDataPart, Part } from '@google/generative-ai';

import {
  BaseMessage,
  BotMessage,
  ChatMessage,
  ContentLike,
  HumanMessage,
} from '../../../../../input/load/msgs/base.js';
import {
  GeminiContentRole,
  GeminiInlineData,
  GeminiMimeType,
} from '../index.js';

export function getMessageFromContentWithRole(
  content: Content,
  defaultRole?: GeminiContentRole
): BaseMessage {
  const role = content.role ?? defaultRole;
  const text: string = content.parts[0].text ?? '';
  const additionalKwargs: Record<string, unknown> = {};

  switch (role) {
    case 'user':
      return new HumanMessage({ content: text });
    case 'model':
      return new BotMessage({ content: text, additionalKwargs });
    default:
      return new ChatMessage({
        content: text,
        role: role ?? 'unknown',
        additionalKwargs,
      });
  }
}

export function getMessageFromContent(content: Content): BaseMessage {
  const text: string = content.parts[0].text ?? '';
  const additionalKwargs: Record<string, unknown> = {};

  switch (content.role) {
    case 'user':
      return new HumanMessage({ content: text });
    case 'model':
      return new BotMessage({ content: text, additionalKwargs });
    default:
      return new ChatMessage({
        content: text,
        role: content.role ?? 'unknown',
        additionalKwargs,
      });
  }
}

export function getContentFromMessage(message: BaseMessage): Content {
  return {
    role: getGeminiRoleFromMessage(message),
    parts: getPartsFromMessage(message),
  };
}

export function getGeminiRoleFromMessage(
  message: BaseMessage
): GeminiContentRole {
  switch (message._role()) {
    case 'system':
    case 'human':
      return 'user';
    case 'assistant':
      return 'model';
    default:
      throw new Error(
        `Message role ${message._role()} does not support Gemini.`
      );
  }
}

export function getPartsFromMessage(message: BaseMessage): Part[] {
  const getPart = (content: ContentLike): Part => {
    if (typeof content === 'string') {
      return {
        text: content,
      };
    } else if (isTextData(content)) {
      return getTextPart(content);
    } else if (isInlineData(content)) {
      return getInlinePartFromInlineData(content);
    }

    throw new Error('Message is not valid for Gemini');
  };

  if (!Array.isArray(message.content)) {
    return [getPart(message.content)];
  }

  return message.content.map((content: ContentLike) => getPart(content));
}

export function isModalityRequiredInMessage(message: BaseMessage): boolean {
  const isModalityRequired = (content: ContentLike): boolean => {
    return typeof content !== 'string' && isInlineData(content);
  };

  if (!Array.isArray(message.content)) {
    return isModalityRequired(message.content);
  }

  return message.content.some((content: ContentLike) =>
    isModalityRequired(content)
  );
}

export function getTextPart(content: any): any {
  return {
    text: content.text,
  };
}

export function getInlinePartFromInlineData(
  inlineData: GeminiInlineData
): InlineDataPart {
  return {
    inlineData: {
      mimeType: inlineData.mimeType,
      data: inlineData.data,
    },
  };
}

export function isTextData(partLike: any): boolean {
  return (
    typeof partLike === 'object' &&
    'text' in partLike &&
    typeof partLike['text'] === 'string' &&
    partLike['text'].length > 0
  );
}

export function isInlineData(partLike: any): partLike is GeminiInlineData {
  return (
    typeof partLike === 'object' &&
    'mimeType' in partLike &&
    'data' in partLike &&
    isValidMime(partLike['mimeType']) &&
    typeof partLike['data'] === 'string'
  );
}

export function isValidMime(mimeLike: any): mimeLike is GeminiMimeType {
  return Object.values(GeminiMimeType).includes(mimeLike as GeminiMimeType);
}
