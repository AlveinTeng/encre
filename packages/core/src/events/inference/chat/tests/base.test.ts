import { expect, test } from '@jest/globals';
import { stringify } from 'yaml';
import { MemoryCache } from '../../../../cache/index.js';
import {
  BaseMessage,
  BaseMessageLike,
} from '../../../input/load/msgs/base.js';
import { ChatMessage } from '../../../input/load/msgs/chat.js';
import { HumanMessage } from '../../../input/load/msgs/human.js';
import { SystemMessage } from '../../../input/load/msgs/system.js';
import { convertMessageLikeToMessage } from '../../../input/load/msgs/utils.js';
import { BasePrompt } from '../../../input/load/prompts/base.js';
import { ChatPrompt } from '../../../input/load/prompts/chat.js';
import { StringPrompt } from '../../../input/load/prompts/text.js';
import { Generation } from '../../../output/provide/generation.js';
import { LLMResult } from '../../../output/provide/llmresult.js';
import {
  BaseChatLM,
  BaseLLM,
  BaseLM,
  BaseLMCallOptions,
  BaseLMInput,
  calculateMaxToken,
} from '../base.js';

test('test calculateMaxToken', async () => {
  const prompt = 'this is a prompt.';

  const modelName = 'gpt-3.5-turbo-16k-0613';
  expect(await calculateMaxToken(prompt, modelName)).toBe(16384 - 5);

  const modelName2 = 'gpt-3.5-turbo-0613';
  expect(await calculateMaxToken(prompt, modelName2)).toBe(4096 - 5);

  const modelName3 = 'gpt-4-32k-0314';
  expect(await calculateMaxToken(prompt, modelName3)).toBe(32768 - 5);

  const modelName4 = 'gpt-4-0314';
  expect(await calculateMaxToken(prompt, modelName4)).toBe(8192 - 5);

  const modelName5 = 'gpt2';
  expect(await calculateMaxToken(prompt, modelName5)).toBe(4096 - 5);
});

test('test BaseLM', async () => {
  class TestLM extends BaseLM {
    _eventNamespace(): string[] {
      return ['tests'];
    }

    _llmType(): string {
      return 'test_llm';
    }

    _modelType(): string {
      return 'test_model';
    }

    /**
     * In this method, we mock the language model result as pairs of prompts
     * given an array of prompts.
     * @param {string} prompt - A prompt.
     * @param options if given an array of strings, then it means the stopwords.
     * @param callbacks TODO: not yet implemented
     */
    async provide(
      prompt: string,
      options?: BaseLMCallOptions | string[],
      callbacks?: any
    ): Promise<LLMResult> {
      const generations: Generation[] = [
        {
          output: prompt,
          info: {
            index: 0,
          },
        },
      ];

      return { generations };
    }

    /**
     * Invokes the language model with a given input and options.
     * @param input The given prompt in {@link BaseLMInput}
     * @param options It contains callbask and other options such as stopwords.
     */
    async invoke(
      input: BaseLMInput,
      options?: Partial<BaseLMCallOptions>
    ): Promise<unknown> {
      let prompt: BasePrompt;
      if (typeof input === 'string') {
        prompt = new StringPrompt(input);
      } else if (Array.isArray(input)) {
        prompt = new ChatPrompt(
          input.map((msg: BaseMessageLike) => convertMessageLikeToMessage(msg))
        );
      } else {
        prompt = input;
      }

      const result: LLMResult = await this.provide(
        prompt.toString(),
        options,
        options?.callbacks
      );

      return result;
    }
  }

  const testLM = new TestLM({});

  // BaseLM is not serializable
  const serializedStr: string = JSON.stringify(testLM, null, 2);
  expect(stringify(JSON.parse(serializedStr))).toMatchSnapshot();

  const result: LLMResult = await testLM.provide('this is a prompt.');
  expect(result).toStrictEqual({
    generations: [{ info: { index: 0 }, output: 'this is a prompt.' }],
  });

  expect(stringify(await testLM.invoke('this is a prompt.'))).toMatchSnapshot();

  expect(
    stringify(
      await testLM.invoke([
        ['system', 'this is a system message'],
        ['human', 'this is a human message'],
        ['assistant', 'this is a bot message'],
      ])
    )
  ).toMatchSnapshot();

  expect(
    stringify(
      await testLM.invoke([
        'this is a human message 1',
        'this is a human message 2',
        'this is a human message 3',
      ])
    )
  ).toMatchSnapshot();

  expect(
    stringify(await testLM.invoke(new StringPrompt('this is a prompt.')))
  ).toMatchSnapshot();

  expect(
    stringify(
      await testLM.invoke(
        new ChatPrompt([
          new SystemMessage('this is a system message'),
          new HumanMessage('this is a human message'),
        ])
      )
    )
  ).toMatchSnapshot();

  expect(
    stringify(
      await testLM.invoke([
        new SystemMessage('this is a system message'),
        new HumanMessage('this is a human message'),
      ])
    )
  ).toMatchSnapshot();
});

test('test BaseLLM', async () => {
  class TestLLM extends BaseLLM {
    getParams(
      options?: this['SerializedCallOptions']
    ): Record<string, unknown> {
      return {};
    }

    /**
     * In this method, we mock the language model result as pairs of prompts
     * given an array of prompts.
     * @param {string} prompt - A prompt.
     * @param options if given an array of strings, then it means the stopwords.
     * @param callbacks TODO: not yet implemented
     */
    async _provide(
      prompt: string,
      options: this['SerializedCallOptions']
    ): Promise<LLMResult> {
      const generations: Generation[] = [
        {
          output: prompt,
          info: {
            index: 0,
          },
        },
      ];

      return { generations };
    }

    _llmType(): string {
      return 'test_llm';
    }
  }

  const cache: MemoryCache = MemoryCache.global();
  const testLLMWithCustomCache = new TestLLM({ cache });

  // BaseLLM is not serializable
  const serializedStr: string = JSON.stringify(testLLMWithCustomCache, null, 2);
  expect(stringify(JSON.parse(serializedStr))).toMatchSnapshot();

  const result: LLMResult =
    await testLLMWithCustomCache.provide('this is a prompt.');
  expect(result).toStrictEqual({
    generations: [{ info: { index: 0 }, output: 'this is a prompt.' }],
    llmOutput: {},
  });

  expect(
    stringify(await testLLMWithCustomCache.invoke('this is a prompt.'))
  ).toMatchSnapshot();

  expect(
    stringify(
      await testLLMWithCustomCache.invoke([
        ['system', 'this is a system message'],
        ['human', 'this is a human message'],
        ['assistant', 'this is a bot message'],
      ])
    )
  ).toMatchSnapshot();

  expect(
    stringify(
      await testLLMWithCustomCache.invoke([
        'this is a human message 1',
        'this is a human message 2',
        'this is a human message 3',
      ])
    )
  ).toMatchSnapshot();

  expect(
    stringify(
      await testLLMWithCustomCache.invoke(new StringPrompt('this is a prompt.'))
    )
  ).toMatchSnapshot();

  expect(
    stringify(
      await testLLMWithCustomCache.invoke(
        new ChatPrompt([
          new SystemMessage('this is a system message'),
          new HumanMessage('this is a human message'),
        ])
      )
    )
  ).toMatchSnapshot();

  expect(
    stringify(
      await testLLMWithCustomCache.invoke([
        new SystemMessage('this is a system message'),
        new HumanMessage('this is a human message'),
      ])
    )
  ).toMatchSnapshot();
});

test('test BaseChatLM', async () => {
  class TestChatLM extends BaseChatLM {
    getParams(
      options?: this['SerializedCallOptions']
    ): Record<string, unknown> {
      return {};
    }

    /**
     * In this method, we mock the language model result as pairs of prompts
     * given an array of prompts.
     * @param {string} prompt - A prompt.
     * @param options if given an array of strings, then it means the stopwords.
     * @param callbacks TODO: not yet implemented
     */
    async _provide(
      messages: BaseMessage[],
      options: this['SerializedCallOptions']
    ): Promise<LLMResult> {
      const generations: Generation[] = messages.map(
        (message: BaseMessage, idx: number) => ({
          output: message.content ?? '',
          info: {
            index: idx,
          },
        })
      );

      return { generations };
    }

    _llmType(): string {
      return 'test_chatlm';
    }
  }

  const cache: MemoryCache = MemoryCache.global();
  const testChatLMWithCustomCache = new TestChatLM({ cache });

  // BaseLLM is not serializable
  const serializedStr: string = JSON.stringify(
    testChatLMWithCustomCache,
    null,
    2
  );
  expect(stringify(JSON.parse(serializedStr))).toMatchSnapshot();

  const result: LLMResult = await testChatLMWithCustomCache.provide([
    new ChatMessage('this is a prompt.', 'assistant'),
  ]);
  expect(result).toStrictEqual({
    generations: [{ info: { index: 0 }, output: 'this is a prompt.' }],
    llmOutput: {},
  });

  expect(
    stringify(await testChatLMWithCustomCache.invoke('this is a prompt.'))
  ).toMatchSnapshot();

  expect(
    stringify(
      await testChatLMWithCustomCache.invoke([
        ['system', 'this is a system message'],
        ['human', 'this is a human message'],
        ['assistant', 'this is a bot message'],
      ])
    )
  ).toMatchSnapshot();

  expect(
    stringify(
      await testChatLMWithCustomCache.invoke([
        'this is a human message 1',
        'this is a human message 2',
        'this is a human message 3',
      ])
    )
  ).toMatchSnapshot();

  expect(
    stringify(
      await testChatLMWithCustomCache.invoke(
        new StringPrompt('this is a prompt.')
      )
    )
  ).toMatchSnapshot();

  expect(
    stringify(
      await testChatLMWithCustomCache.invoke(
        new ChatPrompt([
          new SystemMessage('this is a system message'),
          new HumanMessage('this is a human message'),
        ])
      )
    )
  ).toMatchSnapshot();

  expect(
    stringify(
      await testChatLMWithCustomCache.invoke([
        new SystemMessage('this is a system message'),
        new HumanMessage('this is a human message'),
      ])
    )
  ).toMatchSnapshot();
});
