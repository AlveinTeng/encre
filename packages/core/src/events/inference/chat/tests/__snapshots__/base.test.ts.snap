// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test BaseChatLM 1`] = `
"_grp: 1
_type: not_implemented
_id:
  - inference
  - chat
  - llms
  - test_chatlm
  - TestChatLM
"
`;

exports[`test BaseChatLM 2`] = `
"generations:
  - output: this is a prompt.
    info:
      index: 0
llmOutput: {}
"
`;

exports[`test BaseChatLM 3`] = `
"generations:
  - output: this is a system message
    info:
      index: 0
  - output: this is a human message
    info:
      index: 1
  - output: this is a bot message
    info:
      index: 2
llmOutput: {}
"
`;

exports[`test BaseChatLM 4`] = `
"generations:
  - output: this is a human message 1
    info:
      index: 0
  - output: this is a human message 2
    info:
      index: 1
  - output: this is a human message 3
    info:
      index: 2
llmOutput: {}
"
`;

exports[`test BaseChatLM 5`] = `
"generations:
  - output: this is a prompt.
    info:
      index: 0
llmOutput: {}
"
`;

exports[`test BaseChatLM 6`] = `
"generations:
  - output: this is a system message
    info:
      index: 0
  - output: this is a human message
    info:
      index: 1
llmOutput: {}
"
`;

exports[`test BaseChatLM 7`] = `
"generations:
  - output: this is a system message
    info:
      index: 0
  - output: this is a human message
    info:
      index: 1
llmOutput: {}
"
`;

exports[`test BaseLLM 1`] = `
"_grp: 1
_type: not_implemented
_id:
  - inference
  - chat
  - llms
  - test_llm
  - TestLLM
"
`;

exports[`test BaseLLM 2`] = `
"generations:
  - output: this is a prompt.
    info:
      index: 0
llmOutput: {}
"
`;

exports[`test BaseLLM 3`] = `
"generations:
  - output: |-
      System: this is a system message
      Human: this is a human message
      AI: this is a bot message
    info:
      index: 0
llmOutput: {}
"
`;

exports[`test BaseLLM 4`] = `
"generations:
  - output: |-
      Human: this is a human message 1
      Human: this is a human message 2
      Human: this is a human message 3
    info:
      index: 0
llmOutput: {}
"
`;

exports[`test BaseLLM 5`] = `
"generations:
  - output: this is a prompt.
    info:
      index: 0
llmOutput: {}
"
`;

exports[`test BaseLLM 6`] = `
"generations:
  - output: |-
      System: this is a system message
      Human: this is a human message
    info:
      index: 0
llmOutput: {}
"
`;

exports[`test BaseLLM 7`] = `
"generations:
  - output: |-
      System: this is a system message
      Human: this is a human message
    info:
      index: 0
llmOutput: {}
"
`;

exports[`test BaseLM 1`] = `
"_grp: 1
_type: not_implemented
_id:
  - tests
  - TestLM
"
`;

exports[`test BaseLM 2`] = `
"generations:
  - output: this is a prompt.
    info:
      index: 0
"
`;

exports[`test BaseLM 3`] = `
"generations:
  - output: |-
      System: this is a system message
      Human: this is a human message
      AI: this is a bot message
    info:
      index: 0
"
`;

exports[`test BaseLM 4`] = `
"generations:
  - output: |-
      Human: this is a human message 1
      Human: this is a human message 2
      Human: this is a human message 3
    info:
      index: 0
"
`;

exports[`test BaseLM 5`] = `
"generations:
  - output: this is a prompt.
    info:
      index: 0
"
`;

exports[`test BaseLM 6`] = `
"generations:
  - output: |-
      System: this is a system message
      Human: this is a human message
    info:
      index: 0
"
`;

exports[`test BaseLM 7`] = `
"generations:
  - output: |-
      System: this is a system message
      Human: this is a human message
    info:
      index: 0
"
`;
